backend/SC4SNMP_UI_backend/apply_changes/handle_changes.py

```diff
@@ -0,0 +1,153 @@
+            # If records isn't currently scheduled, schedule it and update its state in mongo
+            mongo_config_collection.update_one({"_id": record["_id"]},
+                                         {"$set": {"currently_scheduled": True}})
+            run_job.apply_async(countdown=request["job_delay"], queue='apply_changes')
+        return request["job_delay"]
+
+
+class SingletonMeta(type):
+    _instances = {}
+    _lock: Lock = Lock()
+
+    def __call__(cls, *args, **kwargs):
+        with cls._lock:
+            if cls not in cls._instances:
+                instance = super().__call__(*args, **kwargs)
+                cls._instances[cls] = instance
+        return cls._instances[cls]
+
+class ApplyChanges(metaclass=SingletonMeta):
+    def __init__(self) -> None:
+        """
+        ApplyChanges is a singleton responsible for creating mongo record with a current state of kubernetes job.
+        """
+        self.__handling_chain = CheckJobHandler()
+        schedule_handler = ScheduleHandler()
+        self.__handling_chain.set_next(schedule_handler)
+        if len(list(mongo_config_collection.find({}))) == 0:
+            # If statement is used here instead of upsert operation
+            # because __init__ method should only create, not update
+            # this record.
+            new_config = {
+                "previous_job_start_time": None,
+                "currently_scheduled": False
+            }
+            mongo_config_collection.insert_one(new_config)
+
+
+    def apply_changes(self):
+        """
+        Run chain of actions to schedule new kubernetes job.
+        """
+        job_delay = self.__handling_chain.handle()
+        return job_delay
+
+@shared_task()
+def run_job():
+    job = None
+    batch_v1 = None
+    with open(JOB_CONFIG_PATH, encoding="utf-8") as file:
+        config_file = yaml.safe_load(file)
+        if config_file["apiVersion"] != "batch/v1":
+            raise ValueError("api version is different from batch/v1")
+        config.load_incluster_config()
+        batch_v1 = client.BatchV1Api()
+        job = create_job_object(config_file)
+
+    with MongoClient(MONGO_URI) as connection:
+        try_creating = True
+        while try_creating:
+            # Try creating new job. If the previous job is still present in the namespace,
+            # ApiException will we be raised. In that happens wait for 10 seconds and try creating the job again
+            try:
+                create_job(batch_v1, job, JOB_NAMESPACE)
+                try_creating = False
+                record = list(connection.sc4snmp.config_collection.find())[0]
+                connection.sc4snmp.config_collection.update_one({"_id": record["_id"]},
+                                             {"$set": {"previous_job_start_time": datetime.datetime.utcnow(),
+                                                       "currently_scheduled": False}})
+            except ApiException:
```

Comment:
I'd add a logger here si user know what's going on, also I'd ad another exception in case Mongo error happens

Reply:
Also, will it fail infinitely? Can we add max retries? At some point I think we should return the ultimate ERROR